ecosystem/indexer-grpc/indexer-grpc-manager/src/metadata_manager.rs

```diff
@@ -0,0 +1,383 @@
+        let staleness = now_since_epoch.saturating_sub(latest_state_timestamp_since_epoch);
+
+        staleness >= threshold
+    }
+
+    pub(crate) async fn start(&self) -> Result<()> {
+        loop {
+            tokio_scoped::scope(|s| {
+                for kv in &self.grpc_managers {
+                    let grpc_manager = kv.value();
+                    let client = grpc_manager.client.clone();
+                    s.spawn(async move {
+                        let _ = self.heartbeat(client).await;
+                    });
+                }
+
+                for kv in &self.fullnodes {
+                    let (address, fullnode) = kv.pair();
+                    let need_ping = fullnode.recent_states.back().map_or(true, |s| {
+                        Self::need_ping(s.timestamp.unwrap_or_default(), Duration::from_secs(1))
+                    });
+                    if need_ping {
+                        let address = address.clone();
+                        let client = fullnode.client.clone();
+                        s.spawn(async move {
+                            let _ = self.ping_fullnode(address, client).await;
+                        });
+                    }
+                }
+
+                for kv in &self.live_data_services {
+                    let (address, live_data_service) = kv.pair();
+                    let need_ping = live_data_service.recent_states.back().map_or(true, |s| {
+                        Self::need_ping(s.timestamp.unwrap_or_default(), Duration::from_secs(5))
+                    });
+                    if need_ping {
+                        let address = address.clone();
+                        let client = live_data_service.client.clone();
+                        s.spawn(async move {
+                            let _ = self.ping_live_data_service(address, client).await;
+                        });
+                    }
+                }
+
+                for kv in &self.historical_data_services {
+                    let (address, historical_data_service) = kv.pair();
+                    let need_ping =
+                        historical_data_service
+                            .recent_states
+                            .back()
+                            .map_or(true, |s| {
+                                Self::need_ping(
+                                    s.timestamp.unwrap_or_default(),
+                                    Duration::from_secs(5),
+                                )
+                            });
+                    if need_ping {
+                        let address = address.clone();
+                        let client = historical_data_service.client.clone();
+                        s.spawn(async move {
+                            let _ = self.ping_historical_data_service(address, client).await;
+                        });
+                    }
+                }
+            });
+
+            tokio::time::sleep(Duration::from_secs(1)).await;
```

Comment:
Why do we need to ping so often? Can we sleep longer here?

Reply:
I actually want to make it more frequent since it's cheap. Under load a lot of things can happen in 1s.